pkg/task/task.go

```diff
@@ -102,39 +108,114 @@ func (pp *Profiler) logProfileSummary(taskName string) {
 
 // Manages Task execution
 type taskRunner struct {
-	task   Task
-	writer filewriter.FileWriter
+	task           Task
+	writer         filewriter.FileWriter
+	withCheckpoint bool
+}
+
+type TaskRunnerOpt func(*taskRunner)
+
+func WithCheckpointFile() TaskRunnerOpt {
+	return func(t *taskRunner) {
+		t.withCheckpoint = true
+		if !features.IsActive(features.CheckpointEnabled()) {
+			logger.V(4).Info("CHECKPOINT_ENABLED env var is not set. Continuing operation without this feature...", features.CheckpointEnabledEnvVar, features.IsActive(features.CheckpointEnabled()))
+			t.withCheckpoint = false
+		}
+	}
 }
 
-// executes Task
 func (pr *taskRunner) RunTask(ctx context.Context, commandContext *CommandContext) error {
+	checkpointFileName := fmt.Sprintf("%s-checkpoint.yaml", commandContext.ClusterSpec.Cluster.Name)
+	var checkpointInfo CheckpointInfo
+	var err error
+
 	commandContext.Profiler = &Profiler{
 		metrics: make(map[string]map[string]time.Duration),
 		starts:  make(map[string]map[string]time.Time),
 	}
 	task := pr.task
 	start := time.Now()
 	defer taskRunnerFinalBlock(start)
+
+	if pr.withCheckpoint {
+		checkpointInfo, err = pr.setupCheckpointInfo(commandContext, checkpointFileName)
+		if err != nil {
+			return err
+		}
+	}
+
 	for task != nil {
+		if pr.withCheckpoint {
+			if completedTask, ok := checkpointInfo.CompletedTasks[task.Name()]; ok {
+				logger.V(4).Info("Restoring task", "task_name", task.Name())
+				nextTask, err := task.Restore(ctx, commandContext, completedTask)
+				if err != nil {
+					return fmt.Errorf("restoring checkpoint info: %v", err)
+				}
+				task = nextTask
+				continue
+			}
+		}
 		logger.V(4).Info("Task start", "task_name", task.Name())
 		commandContext.Profiler.SetStartTask(task.Name())
 		nextTask := task.Run(ctx, commandContext)
 		commandContext.Profiler.MarkDoneTask(task.Name())
 		commandContext.Profiler.logProfileSummary(task.Name())
+		if commandContext.OriginalError == nil && pr.withCheckpoint {
```

Comment:
same here, can't we execute this even if checkpoint is disabled?
For me, checkpoint disabled just means we don't read the checkpoint file at the beginning.

Reply:
Ahh interesting
IMO the task runner is already fully aware of the checkpoint logic. If this is a concern, we can address it and move it somewhere else.

Otherwise, I would vote to remove the check. I think the only thing we should avoid doing when `withCheckpoint` is false is to restore the checkpoint info from the file at the beginning. The rest can stay the same, even writing to file.